### ISSUE TITLE
Annotation for method inlining

### ISSUE THREAD

**Issue by leisim** [dart-lang/sdk#40522]
For method inlining I have found the `@pragma('vm:prefer-inline')`
and `@pragma('dart2js:tryInline')` annotations.

It would be convenient if there was an annotation to suggest inlining to both, the VM and dart2js at the same time. 

There is a [`@tryInline`](https://pub.dev/documentation/meta/latest/meta_dart2js/tryInline-constant.html) annotation in the meta package but it only works for dart2js.

Currently I have to write the following:
```dart
@pragma('vm:prefer-inline')
@pragma('dart2js:tryInline')
void methodToBeInlined() {
 // code
}
```

Something like this would be nice:
```dart
@inline
void methodToBeInlined() {
 // code
}
```

**Comment by vsmenon** [comment]
@mraleph @sigmundch - this would indeed be nice.  :-)

**Comment by mraleph** [comment]
dart2js and Dart VM AOT are rather different beasts so I would say providing the annotation that works for both would do more harm than good. You are only supposed to use this annotation when you really understand the implications and have evaluated the effect of putting it on the function. 

@leisim I would actually be quite curious to know what you are using these annotations for.

**Comment by mraleph** [comment]
@leisim I see you are using it here https://github.com/hivedb/hive/commit/40c700382ee2261c43add8c27d02c4b085f445bf. 

Do you make these inlining decisions based on performance of this benchmark: https://github.com/hivedb/hive_benchmark?

/cc @mkustermann @sstrickl - we should look at this as part of our inlining heuristics tuning project.

**Comment by leisim** [comment]
> You are only supposed to use this annotation when you really understand the implications and have evaluated the effect of putting it on the function

I created a few microbenchmarks and from what I understand, these annotations do almost exactly what you expect them to do. In the generated dart2js code, the function call is replaced with the body of the to-be-inlined function. In the VM, I could not verify that but the benchmarks imply that it works in a similar fashion.

Unfortuantely neither dart2js nor the VM seem to be able to inline async functions:

```dart
Future<int> functionToBeInlined() async {
  var result = await someFuture();
  return result + 5;
}

void main() async {
  print(await functionToBeInlined());
}
```

I would expect the compiled code to be something like this:

```dart
void main() async {
  var result = await someFuture();
  print(result + 5);
}
```

> Do you make these inlining decisions based on performance of this benchmark

I created a few microbenchmarks to verify performance improvements. I use inlining on all very small methods which are used in a performance-critical way (for example [here](https://github.com/hivedb/hive/blob/6822ed1e440bdc912d880da4c2885a68c5f5372d/lib/src/util/extensions.dart#L16)). I could not find any downside to inlining when the method is very small or called only once. It would be very helpful if the documentation was a bit more detailed and would provide best practices for inlining.

**Comment by mraleph** [comment]
> In the generated dart2js code, the function call is replaced with the body of the to-be-inlined function. In the VM, I could not verify that but the benchmarks imply that it works in a similar fashion.

That is correct understanding.  One important metric to watch for is code size (in addition to performance). 

> I would expect the compiled code to be something like this:

Yeah, at least in VM async functions are desugared into something more complicated so normal inlining does not work for them. I have filed https://github.com/dart-lang/sdk/issues/40580 to track this.

> I created microbenchmarks to verify performance improvements. 

Could you publish those microbenchmarks as well? That would help a lot. 

> I could not find any downsides to inlining when the method is small or called only once.

Yeah, that's correct rule of thumb.

**Comment by Afsar-Pasha** [comment]
If multiple annotations can be supported with List, we could write something like,
```dart
const inline = [pragma('vm:prefer-inline'), pragma('dart2js:tryInline')];

@inline
void inlineFunc() {}
```
